core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -284,12 +289,18 @@ class KafkaController(val config : KafkaConfig, zkUtils: ZkUtils, val brokerStat
                   }
                   // If the broker is a follower, updates the isr in ZK and notifies the current leader
                   replicaStateMachine.handleStateChanges(Set(PartitionAndReplica(topicAndPartition.topic,
-                    topicAndPartition.partition, id)), OfflineReplica)
+                    topicAndPartition.partition, id)), OfflineReplica, newBatch = false)
                 }
               }
             }
           }
       }
+
+      inLock(controllerContext.controllerLock) {
```

Comment:
@junrao Yes, I think it is at step (a). You are right that controller doesn't need to wait for those requests to be sent to broker. But broker needs to take those 3000 requests out of queue before it can read ControlledShutdownResponse and finish controlled shutdown.. And taking 3000 requests out of queue can take much longer time than taking only 3 requests out of the queue.

I am thinking that RequestSendThread may split that into multiple requests in the following scenario. After broker puts first 500 LeaderAndIsrRequests into the queue,  RequestSendThread finds itself ready, takes those 500 LeaderAndIsrRequests out of queue, merge them into 1 LeaderAndIsrRequests, and send it to broker. After controller receives response, maybe only 500 more requests are put into the queue, so broker takes them out and send them as 1 request. This goes on and controller will send 6 LeaderAndIsrRequests instead of 1. How efficient it is will depend on how fast controller puts LeaderAndIsrRequests (and others) into the queue and how fast broker replies to those requests from controller.